theano/gpuarray/reduction.py

```diff
@@ -110,80 +81,68 @@ def c_code(self, node, name, input_names, output_names, sub):
         # Recall: fail = sub['fail']
         max_typecode = pygpu.gpuarray.dtype_to_typecode(node.inputs[0].dtype)
         argmax_typecode = pygpu.gpuarray.dtype_to_typecode(self.argmax_dtype)
-        # axes_ctype = pygpu.gpuarray.dtype_to_ctype(node.inputs[1].dtype)
         axes_ctype = 'int64_t'
+        assert node.inputs[1].ndim == 1
         ret = """
+        GpuArray temp;
         GpuArray* %(name)s_input = &%(X)s->ga;
         size_t %(name)s_input_ndim = PyGpuArray_NDIM(%(X)s);
-        """
-        if NoneConst.equals(node.inputs[1]):
-            ret += """
-            unsigned  %(name)s_redux_len = %(name)s_input_ndim;
-            unsigned* %(name)s_axes_to_reduce = new unsigned[%(name)s_redux_len];
-            for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
-                %(name)s_axes_to_reduce[i] = i;
-            }
-            """
-        else:
-            assert node.inputs[1].ndim == 1
-            ret += """
-            unsigned  %(name)s_redux_len = PyArray_DIM(%(axes)s, 0);
-            unsigned* %(name)s_axes_to_reduce = new unsigned[%(name)s_redux_len];
-            for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
-                %(name)s_axes_to_reduce[i] =
-                    (unsigned)( ((%(axes_ctype)s*)PyArray_DATA(%(axes)s)) [i * (PyArray_STRIDES(%(axes)s)[0] / sizeof(%(axes_ctype)s))] );
-            }
-            """
-        ret += """
+
+        unsigned  %(name)s_redux_len = PyArray_DIM(%(axes)s, 0);
+        unsigned* %(name)s_axes_to_reduce = (unsigned*)malloc(%(name)s_redux_len * sizeof(unsigned));
+        for (unsigned i = 0; i < %(name)s_redux_len; ++i) {
+            %(name)s_axes_to_reduce[i] = (unsigned) (*(%(axes_ctype)s*)PyArray_GETPTR1(%(axes)s, i));
+        }
+
         size_t  %(name)s_output_ndim = %(name)s_input_ndim - %(name)s_redux_len;
         size_t* %(name)s_output_dims = NULL;
-        if(%(name)s_output_ndim == 0) {
+        if (%(name)s_output_ndim == 0) {
             /* Current backend function GpuArray_maxandargmax does not work when
              * all axes need to be reduced. So to handle this case, we create a view
              * of the input as a matrix with 1 row and as many columns as elements
              * in the input, so that the 2nd dimenson of the matrix will be reduced. */
```

Comment:
One of the problems here is that you call `%(fail)s` without ensuring that a python exception is set.  The second one is that it does not seem to work.

We should fix both.  I would prefer a proper fix for the issue of all-reduced rather than a workaround like this.